website/blog/2023-09-14-ferretdb-v1-10-production-ready-sqlite.md

```diff
@@ -0,0 +1,99 @@
+
+![Old MangoDB architecture](/img/blog/new-arch/image1.png)
+
+The protocol handling module was (and still is) responsible for handling client connections, MongoDB wire protocol,
+and BSON documents encoding/decoding.
+The commands handling module was responsible for handling protocol commands such as `find`, `update`, and `delete`.
+
+Over time, we understood that there is some common code between command handlers.
+We extracted a part responsible for accessing PostgreSQL and running queries into the backend package:
+
+![Old FerretDB architecture](/img/blog/new-arch/image2.png)
+
+When we added another backend, we could not figure out if there was a common interface for different storages as they were so different.
+So we decided to make another commands handling module that was similar in things like command parameters extraction
+but very different in the way it interacted with storage.
+Similar parts were extracted into a set of common packages:
+
+![FerretDB with Tigris architecture](/img/blog/new-arch/image3.png)
+
+Over time the common part grew, but we still had very different backend packages.
+When we decided to add SQLite support, we wanted to revisit our previous decisions –
+could we design a common storage interface after all?
+We did not expect many people to use SQLite with a lot of data, so performance could have a bit lower priority,
+and the interface could be relatively simple.
+At the same time, we had a lot of practical knowledge of how FerretDB is used with PostgreSQL backend,
+what is important and what is not, and what problems the current backend has and how to avoid them.
+So we settled on this architecture:
+
+![FerretDB with SQLite architecture](/img/blog/new-arch/image4.png)
+
+The common backend interface is relatively small and makes it easy to add new backends.
+For example, there is a single method that backends should implement to insert given documents.
+The code for processing ordered and unordered insertions is written once in the command handler;
+there is no need to repeat that logic for every backend.
+Of course, if there is some storage that could perform this operation more efficiently,
+there could be a separate optional method that could be implemented.
+Then the command handler could check if that method is implemented by the current backend and use it.
+
+That architecture also allows us to add optional functionality to all backends at once.
+For example, there is already a _very early_ prototype of the "OpLog" functionality.
+It works by wrapping `insert`, `update`, and `delete` methods of any backend and inserting documents
+into operation log collection when wrapped methods succeed.
+The more advanced backend-specific methods of implementing that functionality could be implemented if needed.
+For example, PostgreSQL could use triggers for achieving better performance.
+That leads us to the discussion of the FerretDB future.
+
+## The future
```

Comment:
The "future" section is unclear to me, with a lot of short, and sometimes confusing sentences.

Suggestion:

With the new backend architecture, it is now easier to contribute to FerretDB, and features can be added faster. This means that moving forward, it will be possible to make FerretDB compatible with MongoDB workloads much easier than before. Also, the resulting code will be easier to maintain as well.

In the coming weeks, we will be migrating our Postgres backend onto the new architecture, and release it without breaking changes. Our aim is to deliver all of these improvements without needing to migrate your data.

Reply:
![image](https://github.com/FerretDB/FerretDB/assets/11512/17f07af8-ed9b-4ac6-9c09-0c67df709bad)